### PR TITLE
{Extension} Install system dependencies for rdbms-conenct extension

### LINKED ISSUE #16250: az extension add -n rdbms-connect fails with PEP 517 issue
> ### `az feedback` auto-generates most of the information requested below, as of CLI version 2.0.62

**Describe the bug**

```
  running build_ext
  building 'setproctitle' extension
  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
  Building wheel for setproctitle (PEP 517): finished with status 'error'
  ERROR: Failed building wheel for setproctitle
  Created temporary directory: C:\Users\zuh\AppData\Local\Temp\pip-wheel-_45dj_px
  Building wheel for terminaltables (setup.py): started
  Destination directory: C:\Users\zuh\AppData\Local\Temp\pip-wheel-_45dj_px
  Running command 'C:\Users\zuh\AppData\Local\Programs\Python\Python37\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\zuh\\AppData\\Local\\Temp\\pip-install-2f6vdk41\\terminaltables_3ebde5834b47490c9b81b1928612da91\\setup.py'"'"'; __file__='"'"'C:\\Users\\zuh\\AppData\\Local\\Temp\\pip-install-2f6vdk41\\terminaltables_3ebde5834b47490c9b81b1928612da91\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\zuh\AppData\Local\Temp\pip-wheel-_45dj_px'
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib
  creating build\lib\terminaltables
  copying terminaltables\ascii_table.py -> build\lib\terminaltables
  copying terminaltables\base_table.py -> build\lib\terminaltables
  copying terminaltables\build.py -> build\lib\terminaltables
  copying terminaltables\github_table.py -> build\lib\terminaltables
  copying terminaltables\other_tables.py -> build\lib\terminaltables
  copying terminaltables\terminal_io.py -> build\lib\terminaltables
  copying terminaltables\width_and_alignment.py -> build\lib\terminaltables
  copying terminaltables\__init__.py -> build\lib\terminaltables
  installing to build\bdist.win-amd64\wheel
  running install
  running install_lib
  creating build\bdist.win-amd64
  creating build\bdist.win-amd64\wheel
  creating build\bdist.win-amd64\wheel\terminaltables
  copying build\lib\terminaltables\ascii_table.py -> build\bdist.win-amd64\wheel\.\terminaltables
  copying build\lib\terminaltables\base_table.py -> build\bdist.win-amd64\wheel\.\terminaltables
  copying build\lib\terminaltables\build.py -> build\bdist.win-amd64\wheel\.\terminaltables
  copying build\lib\terminaltables\github_table.py -> build\bdist.win-amd64\wheel\.\terminaltables
  copying build\lib\terminaltables\other_tables.py -> build\bdist.win-amd64\wheel\.\terminaltables
  copying build\lib\terminaltables\terminal_io.py -> build\bdist.win-amd64\wheel\.\terminaltables
  copying build\lib\terminaltables\width_and_alignment.py -> build\bdist.win-amd64\wheel\.\terminaltables
  copying build\lib\terminaltables\__init__.py -> build\bdist.win-amd64\wheel\.\terminaltables
  running install_egg_info
  running egg_info
  writing terminaltables.egg-info\PKG-INFO
  writing dependency_links to terminaltables.egg-info\dependency_links.txt
  writing top-level names to terminaltables.egg-info\top_level.txt
  reading manifest file 'terminaltables.egg-info\SOURCES.txt'
  writing manifest file 'terminaltables.egg-info\SOURCES.txt'
  Copying terminaltables.egg-info to build\bdist.win-amd64\wheel\.\terminaltables-3.1.0-py3.7.egg-info
  running install_scripts
  creating build\bdist.win-amd64\wheel\terminaltables-3.1.0.dist-info\WHEEL
  creating 'C:\Users\zuh\AppData\Local\Temp\pip-wheel-_45dj_px\terminaltables-3.1.0-py3-none-any.whl' and adding 'build\bdist.win-amd64\wheel' to it
  adding 'terminaltables/__init__.py'
  adding 'terminaltables/ascii_table.py'
  adding 'terminaltables/base_table.py'
  adding 'terminaltables/build.py'
  adding 'terminaltables/github_table.py'
  adding 'terminaltables/other_tables.py'
  adding 'terminaltables/terminal_io.py'
  adding 'terminaltables/width_and_alignment.py'
  adding 'terminaltables-3.1.0.dist-info/METADATA'
  adding 'terminaltables-3.1.0.dist-info/WHEEL'
  adding 'terminaltables-3.1.0.dist-info/top_level.txt'
  adding 'terminaltables-3.1.0.dist-info/zip-safe'
  adding 'terminaltables-3.1.0.dist-info/RECORD'
  removing build\bdist.win-amd64\wheel
  Building wheel for terminaltables (setup.py): finished with status 'done'
  Created wheel for terminaltables: filename=terminaltables-3.1.0-py3-none-any.whl size=15353 sha256=836a0da0ce77b3fa4d886b306a2a592798f91a0465f7eaf00dc6d051f23cb680
  Stored in directory: C:\Users\zuh\AppData\Local\Temp\pip-ephem-wheel-cache-2qda9w2w\wheels\ba\ad\c8\2d98360791161cd3db6daf6b5e730f34021fc9367d5879f497
Successfully built configobj terminaltables
Failed to build setproctitle
ERROR: Could not build wheels for setproctitle which use PEP 517 and cannot be installed directly
Exception information:
Traceback (most recent call last):
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\pip\_internal\cli\base_command.py", line 224, in _main
    status = self.run(options, args)
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\pip\_internal\cli\req_command.py", line 180, in wrapper
    return func(self, options, args)
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\pip\_internal\commands\install.py", line 364, in run
    ", ".join(pep517_build_failure_names)
pip._internal.exceptions.InstallationError: Could not build wheels for setproctitle which use PEP 517 and cannot be installed directly
Removed build tracker: 'C:\\Users\\zuh\\AppData\\Local\\Temp\\pip-req-tracker-fa9ox_yq'

Command '['C:\\Users\\zuh\\AppData\\Local\\Programs\\Python\\Python37\\python.exe', '-m', 'pip', 'install', '--target', 'C:\\Users\\zuh\\.azure\\cliextensions\\rdbms-connect', 'C:\\Users\\zuh\\AppData\\Local\\Temp\\tmped5xy2eq\\rdbms_connect-0.1.0-py2.py3-none-any.whl', '-vv', '--disable-pip-version-check', '--no-cache-dir']' returned non-zero exit status 1.
Pip failed so deleting anything we might have installed at C:\Users\zuh\.azure\cliextensions\rdbms-connect
azure.cli.core.util.handle_exception is called with an exception:
Traceback (most recent call last):
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\knack\cli.py", line 215, in invoke
    cmd_result = self.invocation.execute(args)
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\azure\cli\core\commands\__init__.py", line 654, in execute
    raise ex
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\azure\cli\core\commands\__init__.py", line 717, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\azure\cli\core\commands\__init__.py", line 710, in _run_job
    six.reraise(*sys.exc_info())
  File "C:\Users\zuh\AppData\Roaming\Python\Python37\site-packages\six.py", line 693, in reraise
    raise value
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\azure\cli\core\commands\__init__.py", line 688, in _run_job
    result = cmd_copy(params)
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\azure\cli\core\commands\__init__.py", line 325, in __call__
    return self.handler(*args, **kwargs)
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\azure\cli\core\__init__.py", line 784, in default_command_handler
    return op(**command_args)
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\azure\cli\command_modules\extension\custom.py", line 18, in add_extension_cmd
    version=version, upgrade=upgrade)
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\azure\cli\core\extension\operations.py", line 247, in add_extension
    pip_extra_index_urls=pip_extra_index_urls, pip_proxy=pip_proxy, system=system)
  File "C:\Users\zuh\AppData\Local\Programs\Python\Python37\lib\site-packages\azure\cli\core\extension\operations.py", line 161, in _add_whl_ext
    'Use --debug for more information.'.format(pip_status_code))
knack.util.CLIError: An error occurred. Pip failed with status code 1. Use --debug for more information.
```
**To Reproduce**

`az extension add -n rdbms-connect `

**Expected behavior**

successfully installed

**Environment summary**

```
❯ az version
{
  "azure-cli": "2.16.0",
  "azure-cli-core": "2.16.0",
  "azure-cli-telemetry": "1.0.6",
```
OS version: Windows10
pip version: pip 20.3.1 from c:\users\zuh\appdata\local\programs\python\python37\lib\site-packages\pip (python 3.7)
**Additional context**

### PR DESCRIPTION
**Pull request by fengzhou-msft**
**Description**

Resolve #16250.
The `rdbms-connect` extension depends on `pgcli` which eventually depends on [psycopg2](https://pypi.org/project/psycopg2/). `psycopg2` does not release wheels for Linux/macOS, it will be built from source code during installation of the `rdbms-connect` extension. The build process needs extra system dependencies. We need to handle it before the extension is downloaded and installed with pip.

Tested with Ubuntu Focal, CentOS 8, Fedora 30, SLES 15 and macOS, windows.

**Testing Guide**

az extension add -s https://azurecliprod.blob.core.windows.net/cli-extensions/rdbms_connect-0.1.0-py2.py3-none-any.whl -y

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by fengzhou-msft** [comment]
@Juliehzl please help review.

**Comment by jiasli** [comment]
Perhaps the same logic can be applied to [using `msal-extension` with PyGObject](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/wiki/Encryption-on-Linux).

**Comment by fengzhou-msft** [comment]
> Perhaps the same logic can be applied to [using `msal-extension` with PyGObject](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/wiki/Encryption-on-Linux).

The usage of PyGObject is also triggered by a command?

**Comment by Juliehzl** [comment]
Add @mjain2 for review.

File: src/azure-cli-core/azure/cli/core/extension/operations.py [hunk 1]
```diff
@@ -167,6 +169,75 @@ def _add_whl_ext(cli_ctx, source, ext_sha256=None, pip_extra_index_urls=None, pi
     return extension_name
 
 
+def _install_deps_for_rdbms_connect():  # pylint: disable=too-many-statements
```
Comment (Juliehzl) [inline src/azure-cli-core/azure/cli/core/extension/operations.py:172]: could we name it as `_install_deps_for_psycopg2`? Then other extensions could reuse it.

**Review by mjain2 — APPROVED** [review]
lgtm from rdbms-connect perspective.  Thanks!

File: src/azure-cli-core/azure/cli/core/extension/operations.py (hunk 1 above)
Comment (fengzhou-msft) [reply src/azure-cli-core/azure/cli/core/extension/operations.py:172]: Make sense. 

In the future, we may even want a more generic method to support more packages that require extra system dependencies.

Comment (fengzhou-msft) [reply src/azure-cli-core/azure/cli/core/extension/operations.py]: Updated.

**Review by Juliehzl — APPROVED** [review]

**Pull request merged**